File: src/ariaflow_server/ucc.py
from __future__ import annotations


def ucc_envelope(
    *,
    target: str,
    observed: bool,
    outcome: str,
    completion: str | None = None,
    reason: str = "aggregate",
    detail: str | None = None,
    commands: list[str] | None = None,
) -> dict[str, object]:
    result: dict[str, object] = {
        "observation": "ok" if observed else "failed",
        "outcome": outcome,
        "reason": reason,
        "target": target,
    }
    if completion is not None:
        result["completion"] = completion
    if detail is not None:
        result["message"] = detail
    if commands is not None:
        result["commands"] = commands
    return {
        "meta": {"contract": "UCC", "version": "2.0", "target": target},
        "result": result,
    }


def ucc_record(
    *,
    target: str,
    observed: bool,
    outcome: str,
    completion: str | None = None,
    reason: str = "aggregate",
    detail: str | None = None,
    commands: list[str] | None = None,
) -> dict[str, object]:
    return ucc_envelope(
        target=target,
        observed=observed,
        outcome=outcome,
        completion=completion,
        reason=reason,
        detail=detail,
        commands=commands,
    )

File: packages/core/src/contracts/ucc.ts
export interface UccEnvelopeInput {
  target: string;
  observed: boolean;
  outcome: string;
  completion?: string | null;
  reason?: string;
  detail?: string | null;
  commands?: string[] | null;
}

export interface UccResult {
  observation: "ok" | "failed";
  outcome: string;
  reason: string;
  target: string;
  completion?: string;
  message?: string;
  commands?: string[];
}

export interface UccEnvelope {
  meta: { contract: "UCC"; version: "2.0"; target: string };
  result: UccResult;
}

/**
 * Build the canonical UCC v2.0 response envelope.
 * Mirrors src/ariaflow_server/ucc.py::ucc_envelope.
 */
export function uccEnvelope(input: UccEnvelopeInput): UccEnvelope {
  const result: UccResult = {
    observation: input.observed ? "ok" : "failed",
    outcome: input.outcome,
    reason: input.reason ?? "aggregate",
    target: input.target,
  };
  if (input.completion !== undefined && input.completion !== null) {
    result.completion = input.completion;
  }
  if (input.detail !== undefined && input.detail !== null) {
    result.message = input.detail;
  }
  if (input.commands !== undefined && input.commands !== null) {
    result.commands = input.commands;
  }
  return {
    meta: { contract: "UCC", version: "2.0", target: input.target },
    result,
  };
}

/** Alias preserved for parity with the Python module. */
export const uccRecord = uccEnvelope;
